### PR TITLE
[0.5.0] Add new features

### PR DESCRIPTION
**Pull request by canattofilipe**
Add the following new features:
- Firmware update automatically (dojot/dojot#1502)
- Publish device status (dojot/dojot#1557)
- Discard black listed values (dojot/dojot#1704 )

### CODE REVIEW

File: src/main/java/org/cpqd/iotagent/ImageDownloader.java [hunk 1]
```diff
@@ -45,12 +47,13 @@ public ImageDownloader(String imageManagerUri, String dataDir,
...
 
-    public String downloadImageAndGenerateUri(String tenant, String imageLabel,
-        String version, int protocol) {
+	public String downloadImageAndGenerateUri(String tenant, String imageLabel, String version, int protocol,
+			String imageId, Map<String, String> queryParams) {
 
         String imageFilename = null;
         try {
-            imageFilename = this.fetchImage(tenant, imageLabel, version);
+			imageFilename = imageId == null ? this.fetchImage(tenant, imageLabel, version)
```
Comment (mprevide) [inline src/main/java/org/cpqd/iotagent/ImageDownloader.java:55]: Fix the indentation

File: src/main/java/org/cpqd/iotagent/ImageDownloader.java [hunk 2]
```diff
@@ -79,6 +82,13 @@ public String downloadImageAndGenerateUri(String tenant, String imageLabel,
                 this.mLogger.error("Unknown protocol: " + protocol);
                 throw new RuntimeException("Failed to generate URI");
         }
+		
+		if (queryParams != null) {
```
Comment (mprevide) [inline src/main/java/org/cpqd/iotagent/ImageDownloader.java:86]: Fix the indentation

File: src/main/java/org/cpqd/iotagent/LwM2MAgent.java [hunk 3]
```diff
@@ -514,6 +527,21 @@ public void registered(Registration registration,
                 requestHandler.CancelAllObservations(controlStructure.registration);
                 observeResources(controlStructure.deviceId, controlStructure.tenant,
                         device.getReadableAttributes(), controlStructure.registration);
+
```
Comment (mprevide) [inline src/main/java/org/cpqd/iotagent/LwM2MAgent.java:530]: Fix indentation...

File: src/main/java/org/cpqd/iotagent/LwM2MAgent.java [hunk 4]
```diff
@@ -535,6 +564,11 @@ public void unregistered(Registration registration,
             logger.debug("device left: " + registration.getEndpoint());
 
             deviceMapper.removeSouthboundAssociation(registration.getEndpoint());
+            
+			DeviceControlStructure deviceControlStructure = deviceMapper
```
Comment (mprevide) [inline src/main/java/org/cpqd/iotagent/LwM2MAgent.java:568]: Fiz indentation

File: src/main/java/org/cpqd/iotagent/LwM2MAgent.java [hunk 5]
```diff
@@ -598,6 +632,12 @@ public void onResponse(Observation observation, Registration registration, Obser
 
             eventHandler.updateAttrs(controlStructure.deviceId,
                 controlStructure.tenant, attrJson, null);
+            
+			
+			if (new AutomaticFirmwareUpdate(deviceJson).applyImage(attrJson)) {
```
Comment (mprevide) [inline src/main/java/org/cpqd/iotagent/LwM2MAgent.java:637]: Fiz indentation

**Review by mprevide — COMMENTED** [review]
Fix  all indentations

**Comment by mprevide** [comment]
Try to use a fork the next time.

**Review by mprevide — APPROVED** [review]

**Review by thiagodpf — APPROVED** [review]

**Pull request merged**